File: Java/src/main/java/QuestionBank/Solution101_200/Solution103.java
package QuestionBank.Solution101_200;

import tool.TreeNode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author windows
 */
public class Solution103 {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        ArrayList<List<Integer>> ans = new ArrayList<>();
        HashMap<Integer, List<Integer>> map = new HashMap<>();
        process(root, 0, map);
        for (int i = 0; i < map.size(); i++) {
            List<Integer> list = map.get(i);
            if (i % 2 != 0) {
                for (int j = 0; j < list.size() / 2; j++) {
                    int temp = list.get(j);
                    list.set(j, list.get(list.size() - 1 - j));
                    list.set(list.size() - 1 - j, temp);
                }
                ans.add(list);
            } else {
                ans.add(list);
            }
        }
        return ans;
    }


    private void process(TreeNode treeNode, int depth, Map<Integer, List<Integer>> map) {
        if (treeNode == null){
            return;
        }
        List<Integer> list = map.getOrDefault(depth, new ArrayList<Integer>());
        list.add(treeNode.val);
        map.put(depth, list);
        process(treeNode.left, depth + 1, map);
        process(treeNode.right, depth+1, map);
    }


    public static void main(String[] args) {
        Solution103 solution103 = new Solution103();
    }
}

File: C++/src/Solution101_200/Solution103.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution103.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
#include <map>
#include <functional>

using namespace std;


//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
    vector<vector<int>> ans;
    map<int, vector<int>> hash;

    function<void(TreeNode*, int)> dfs = [&](TreeNode* node, int depth){
        if (node == nullptr){
            return;
        }
        hash[depth].push_back(node->val);
        dfs(node->left, depth + 1);
        dfs(node->right, depth + 1);
    };
    dfs(root, 0);
    int index = 0;
    for (auto &item: hash){
        if (index++ % 2 != 0){
            reverse(item.second.begin(), item.second.end());
        }
        ans.push_back(item.second);
    }
    return ans;
}